iOS-TvShows/iOS-TvShows/Modules/Login/LoginInteractor.swift

```diff
@@ -0,0 +1,67 @@
+//
+//  LoginInteractor.swift
+//  iOS-TvShows
+//
+//  Created by Filip Ban on 08.11.2021..
+//  Copyright (c) 2021 ___ORGANIZATIONNAME___. All rights reserved.
+//
+//  This file was generated by the 🐍 VIPER generator
+//
+
+import Foundation
+import RxSwift
+
+final class LoginInteractor {
+
+    private let service: APIService
+    private let sessionManager: SessionManager
+    private let authStorage: AuthStorageInterface
+
+    init(
+        service: APIService = APIService.instance,
+        sessionManager: SessionManager = .default,
+        authStorage: AuthStorageInterface = AuthStorage.instance
+    ) {
+        self.service = service
+        self.sessionManager = sessionManager
+        self.authStorage = authStorage
+    }
+}
+
+// MARK: - Extensions -
+
+extension LoginInteractor: LoginInteractorInterface {
+
+    func login(email: String, password: String, shouldRemember: Bool) -> Completable {
+        handle(router: LoginRouter.login(email: email, password: password), shouldRemember)
+    }
+
+    func register(email: String, password: String, shouldRemember: Bool) -> Completable {
+        handle(router: LoginRouter.register(email: email, password: password), shouldRemember)
+    }
+
+    private func handle(router: Routable,_ shouldRemember: Bool) -> Completable {
+        service.rx
+            .requestResponse(
+                Data.self,
```

Comment:
How come you are using Data? Not sure if this API call will success since you are not receiving Data, but Codable object. Small hint:

```swift
import Foundation

struct UserResponse: Codable {
    let user: User
}

struct User: Codable {
    let email: String
    let imageUrl: String?
    let id: String

    enum CodingKeys: String, CodingKey {
        case email
        case imageUrl = "image_url"
        case id
    }
}
```

Reply:
.requestResponse(UserResponse.self...

Reply:
Also avoid usage of `keyPath`, try to think why and we can sync today ;)

Reply:
Check [THIS](https://github.com/BanFilip/iOS-TvShows/pull/2/files#diff-afa77d4811b187a3f962c27c9b1de01550321ee7308cdcf73164083aa53ecc6eR34-R46) approach, not liking it... Open for suggestions

Reply:
I would leave method `requestResponse` just like it was before, in case you'll have to parse object AND use headers. Here I would add something like this (but dummy parsing UserResponse is also just fine, even if you won't use it)

```swift
// APIService.swift

    @discardableResult
    open func requestResponse<T: Decodable>(
        _: T.Type,
        keyPath: String? = nil,
        decoder: JSONDecoder = JSONDecoder(),
        router: Routable,
        session: Session,
        completion: @escaping (AFDataResponse<T>) -> Void
    ) -> DataRequest {
        return prepareRequest(for: router, session: session)
            .responseDecodable(keyPath: keyPath, decoder: decoder) { completion($0) }
    }

    @discardableResult
    open func requestResponse(
        router: Routable,
        session: Session,
        completion: @escaping (AFDataResponse<Data?>) -> Void
    ) -> DataRequest {
        return prepareRequest(for: router, session: session)
            .response(completionHandler: completion)
    }
```

```swift
// APIService+Rx.swift

    func requestResponse<T: Decodable>(
        _: T.Type,
        keyPath: String? = nil,
        decoder: JSONDecoder = JSONDecoder(),
        router: Routable,
        session: Session
    ) -> Single<DataResponse<T, AFError>> {
        return Single<DataResponse<T, AFError>> .create { [weak base] single -> Disposable in
            let request = base?.requestResponse(
                T.self,
                keyPath: keyPath,
                decoder: decoder,
                router: router,
                session: session,
                completion: { single(.success($0)) }
            )
            return Disposables.create { request?.cancel() }
        }
    }

    func requestResponse(
        router: Routable,
        session: Session
    ) -> Single<AFDataResponse<Data?>> {
        return Single<AFDataResponse<Data?>> .create { [weak base] single -> Disposable in
            let request = base?.requestResponse(
                router: router,
                session: session,
                completion: { single(.success($0)) }
            )
            return Disposables.create { request?.cancel() }
        }
    }
```